src/Lib.P2P.Tests/PubSub/FloodRouterTest.cs

```diff
@@ -0,0 +1,326 @@
+                    await Task.Delay(100);
+                    peers = (await ns2.PeersAsync(topic)).ToArray();
+                }
+            }
+            finally
+            {
+                await swarm1.StopAsync();
+                await ns1.StopAsync();
+
+                await swarm2.StopAsync();
+                await ns2.StopAsync();
+            }
+        }
+
+        [TestMethod]
+        public async Task Relays_PublishedMessage()
+        {
+            var topic = Guid.NewGuid().ToString();
+
+            var swarm1 = new SwarmService {LocalPeer = self};
+            var router1 = new FloodRouter {SwarmService = swarm1};
+            var ns1 = new PubSubService {LocalPeer = self};
+            ns1.Routers.Add(router1);
+            await swarm1.StartAsync();
+            await ns1.StartAsync();
+
+            var swarm2 = new SwarmService {LocalPeer = other};
+            var router2 = new FloodRouter {SwarmService = swarm2};
+            var ns2 = new PubSubService {LocalPeer = other};
+            ns2.Routers.Add(router2);
+            await swarm2.StartAsync();
+            await ns2.StartAsync();
+
+            var swarm3 = new SwarmService {LocalPeer = other1};
+            var router3 = new FloodRouter {SwarmService = swarm3};
+            var ns3 = new PubSubService {LocalPeer = other1};
+            ns3.Routers.Add(router3);
+            await swarm3.StartAsync();
+            await ns3.StartAsync();
+
+            try
+            {
+                IPublishedMessage lastMessage2 = null;
+                IPublishedMessage lastMessage3 = null;
+                await swarm1.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+                await swarm2.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+                await swarm3.StartListeningAsync("/ip4/127.0.0.1/tcp/0");
+
+                var cs = new CancellationTokenSource();
+                await ns2.SubscribeAsync(topic, msg => lastMessage2 = msg, cs.Token);
+                await ns3.SubscribeAsync(topic, msg => lastMessage3 = msg, cs.Token);
+                await swarm1.ConnectAsync(other);
+                await swarm3.ConnectAsync(other);
+
+                var peers = new Peer[0];
+                var endTime = DateTime.Now.AddSeconds(3);
+                while (peers.Length == 0)
+                {
+                    if (DateTime.Now > endTime)
+                        Assert.Fail("timeout");
+                    await Task.Delay(100);
+                    peers = (await ns2.PeersAsync(topic)).ToArray();
+                }
+
+                CollectionAssert.Contains(peers, other1);
+
+                await ns1.PublishAsync(topic, new byte[] {1});
+                endTime = DateTime.Now.AddSeconds(3);
+                while (lastMessage2 == null || lastMessage3 == null)
+                {
+                    if (DateTime.Now > endTime)
+                        Assert.Fail("timeout");
+                    await Task.Delay(100);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add '.ConfigureAwait(false)' to this call to allow execution to continue in any thread.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4832783)